### PR TITLE
Add stream peer servers limit

### PR DESCRIPTION
**Pull request by janos**
This PR addresses the issue https://github.com/ethersphere/go-ethereum/issues/928 by allowing to limit the number of stream servers per peer.

This PR also contains a small fix in stream package to remove the client from peer map in removeClient function, analogous to the removeServer behaviour.

### CODE REVIEW

File: cmd/swarm/config.go [hunk 1]
```diff
@@ -304,6 +308,12 @@ func envVarsOverride(currentConfig *bzzapi.Config) (config *bzzapi.Config) {
 		}
 	}
 
+	if max := os.Getenv(SWARM_ENV_MAX_STREAM_PEER_SERVERS); max != "" {
+		if m, err := strconv.Atoi(max); err == nil {
```
Comment (zelig) [inline cmd/swarm/config.go]: should we not fatal if `Atoi` fails?

File: swarm/network/stream/peer.go [hunk 2]
```diff
@@ -346,6 +356,7 @@ func (p *Peer) removeClient(s Stream) error {
 		return newNotFoundError("client", s)
 	}
 	client.close()
+	delete(p.clients, s)
```
Comment (zelig) [inline swarm/network/stream/peer.go]: nice catch. this was leaking

File: swarm/network/stream/streamer_test.go [hunk 3]
```diff
@@ -685,3 +686,165 @@ func TestStreamerRequestSubscriptionQuitMsgExchange(t *testing.T) {
...
+		})
+
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
+// TestMaxPeerServersWithoutUnsubscribe creates a registry with a limited
+// number of stream servers, and performs subscriptions to detect sunscriptions
```
Comment (zelig) [inline swarm/network/stream/streamer_test.go]: sunscription -> subscription

File: cmd/swarm/main.go [hunk 4]
```diff
@@ -116,6 +116,12 @@ var (
 		Usage:  "Duration for sync subscriptions update after no new peers are added (default 15s)",
 		EnvVar: SWARM_ENV_SYNC_UPDATE_DELAY,
 	}
+	SwarmMaxStreamPeerServersFlag = cli.IntFlag{
+		Name:   "max-stream-peer-servers",
+		Usage:  "Limit of Stream peer servers, 0 denotes unlimited",
+		EnvVar: SWARM_ENV_MAX_STREAM_PEER_SERVERS,
+		Value:  10000, // A very large default value is possible as stream servers have very small memory footprint
```
Comment (zelig) [inline cmd/swarm/main.go]: small memory footprint by themselves, but if they are actually syncing servers or other IO int . ensive ones, then we need a stricter limit.

File: cmd/swarm/config.go (hunk 1 above)
Comment (janos) [reply cmd/swarm/config.go]: Function envVarsOverride is not returning errors, but we can os.Exit on error in it, as it is a cmd package. Now I see that in this function on other places, we are setting values only if there is error, which is not right:

```
		if swap, err := strconv.ParseBool(swapenable); err != nil {
			currentConfig.SwapEnabled = swap
		}
```

Should we os.Exit on all parsing errors in this function instead?

File: cmd/swarm/main.go (hunk 4 above)
Comment (janos) [reply cmd/swarm/main.go]: Correct, we should evaluate this default value against that also.

File: cmd/swarm/config.go (hunk 1 above)
Comment (zelig) [reply cmd/swarm/config.go]: ooops

Comment (janos) [reply cmd/swarm/config.go]: I have updated the handling for this particular environment variable, and will create an issue for other cases, to be solved in another pr.

**Comment by nonsense** [comment]
@zelig please squash and merge PRs, rather than just merge, as this is messing up the history of the master branch.

**Pull request merged**